Add a `FunctionalForm` class to model functional forms

bradking opened the pull request:
This adds a functional form representation suitable for use in issue #1666.  Basically `FunctionalForm` is a class that can be used as the scalar type of an expression and it records the form of the expression.  See `drake/core/functional_form.h` doxygen comments for detailed documentation.

Model a functional form as a set of function input variables along with an abstract description of how the variables are combined.  We do not need to represent any details of how individual variables appear in the combination.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2322) &emsp; Multiple assignees:&emsp;<img alt="@david-german-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17437069?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/david-german-tri">david-german-tri</a>,&emsp;<img alt="@ggould-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17596413?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/ggould-tri">ggould-tri</a>
<!-- Reviewable:end -->

ggould-tri commented:
+@ggould-tri 

<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> #### Add a \`FunctionalForm\` class to model functional forms
> 
> This adds a functional form representation suitable for use in issue #1666.  Basically `FunctionalForm` is a class that can be used as the scalar type of an expression and it records the form of the expression.  See `drake/core/functional_form.h` doxygen comments for detailed documentation.
> 
> 
> 
> Model a functional form as a set of function input variables along with an abstract description of how the variables are combined.  We do not need to represent any details of how individual variables appear in the combination.
> 
> 
> </details>


---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KHWGJDzKAF8tffmcl5R:-1964593142)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Am I correct that the `!defined(DRAKE_DOXYGEN_CXX)` segments are to prevent doxygen from reaching those sections, and that `DRAKE_DOXYGEN_CXX` is never defined for any real compilation?

<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> +@ggould-tri </details>


---

Reviewed 1 of 5 files at r1.
Review status: 1 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KHWSo0v5JFzmfXXGJBS:515025580)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
From a first glance, this is hugely impressive work and closely matches what we need.  It will be at least a day before I can give it the full review that it deserves.

A couple of initial comments before I wrap up for today, though.

<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> Am I correct that the `!defined(DRAKE_DOXYGEN_CXX)` segments are to prevent doxygen from reaching those sections, and that `DRAKE_DOXYGEN_CXX` is never defined for any real compilation?</details>


---

Reviewed 1 of 5 files at r1.
Review status: 2 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/core/functional_form.cc, line 24 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHWV00Og-18mZ5OuPhK:-KHWV00PlgMWwPys-cSO:1850713579) ([raw file](https://github.com/robotlocomotion/drake/blob/a2b288342742580790eae60e61b7fa6629819cca/drake/core/functional_form.cc#L24)):*
> ```cpp
> class FunctionalForm::Internal {
>  public:
>   static Class const kAdd[8][8];
> ```

Extract this repeated `8` into a constant.  Consider setting the constant to `static_cast<int>(kUndefined)+1` rather than explicit `8` (on the assumption that new enum values will be added before `kUndefined`).

---

*[drake/core/functional_form.cc, line 500 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHWXYYkZkPzdYgOrenR:-KHWXYYlqynGQQ6eRtc5:-1227071658) ([raw file](https://github.com/robotlocomotion/drake/blob/a2b288342742580790eae60e61b7fa6629819cca/drake/core/functional_form.cc#L500)):*
> ```cpp
> //---------------------------------------------------------------------------
> 
> // Define form operation tables using macros for readability.
> ```

Is there a reason to use macros here instead of constants?

---

*[drake/core/functional_form.cc, line 513 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHWXgt0XSnYAx7STNwa:-KHWXgt0XSnYAx7STNwb:-942328508) ([raw file](https://github.com/robotlocomotion/drake/blob/a2b288342742580790eae60e61b7fa6629819cca/drake/core/functional_form.cc#L513)):*
> ```cpp
> FunctionalForm::Class const FunctionalForm::Internal::kAdd[8][8] = {
>     /* clang-format off */
>     /*    \ r:  ZERO  CONS   LIN   AFF  POLY  DIFF   ARB  UNDF */
> ```

This table formatting is a very clever readability solution!

---

*[drake/core/functional_form.h, line 33 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHWUdWO6Imu8kGRVbWn:-KHWUdWO6Imu8kGRVbWo:-673219235) ([raw file](https://github.com/robotlocomotion/drake/blob/a2b288342742580790eae60e61b7fa6629819cca/drake/core/functional_form.h#L33)):*
> ```objc
>  * - \anchor polynomial
>  *   A \b polynomial combination of one or more variables
>  *   (<tt>x'*A*x + b'*x + c</tt>).
> ```

This expression only goes up to quadratic.  You probably need a "..." somewhere in that example.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KHWXDIWuprSpNSkqdTX:1336186381)*
<!-- Sent from Reviewable.io -->

bradking commented:
@ggould-tri Thanks.  I've pushed a revised version of the topic that addresses all the comments so far.

> Am I correct that the !defined(DRAKE_DOXYGEN_CXX) segments are to prevent doxygen from reaching those sections, and that DRAKE_DOXYGEN_CXX is never defined for any real compilation?

Yes.

bradking commented:
I've pushed a revised version of the topic to address `cpplint` diagnostics, including some false positives.

RussTedrake commented:
Reviewed 3 of 5 files at r1, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/core/functional_form.h, line 75 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHdbrUZS1yr7JigVswM:-KHdbrU_nDswgrxjTb9-:-1338211419) ([raw file](https://github.com/robotlocomotion/drake/blob/63c17cf988c78bc89e2d1e0242248d2bb8cb993d/drake/core/functional_form.h#L75)):*
> ```objc
>  *   std::cout << f(x, y) << "\n"; // prints "lin(x,y)"
>  *   std::cout << f(x, c) << "\n"; // prints "aff(x)"
>  *   std::cout << g(2, 3) << "\n"; // prints "6"
> ```

it looks like it is also true that f(x,3) would print aff(x).  that might be nice to add to this doc.

---

*[drake/core/functional_form.h, line 304 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHdcSHYtkF9niCqcIdu:-KHdcSHYtkF9niCqcIdv:842192829) ([raw file](https://github.com/robotlocomotion/drake/blob/63c17cf988c78bc89e2d1e0242248d2bb8cb993d/drake/core/functional_form.h#L304)):*
> ```objc
>    * The contained set is immutable.
>    */
>   class DRAKECORE_EXPORT Variables {
> ```

not required for this PR, but we should coordinate with @ggould-tri to decide if this functionality should be merged with the DecisionVariable logic in the constraint classes.

---

*[drake/core/functional_form.h, line 508 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHdcnBglAQS5u5TyJ06:-KHdcnBglAQS5u5TyJ07:969363924) ([raw file](https://github.com/robotlocomotion/drake/blob/63c17cf988c78bc89e2d1e0242248d2bb8cb993d/drake/core/functional_form.h#L508)):*
> ```objc
>  * Return a copy of \p lhs updated to record addition of a matrix of
>  * \ref constant and/or \ref zero components.
>  */
> ```

very nice. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
@david-german-tri -- it would also be very good to think about how this would be used to label the system methods (e.g. dynamics and output) as being a particular form w/ respect to their inputs.  For instance, can we finally get rid of the isDirectFeedthrough and isTimeInvariant methods and replace them with a FunctionalForm label which contains that same information?  (I especially like that it could be either hard-coded by the author or computed automatically via templates).

<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> I've pushed a revised version of the topic to address `cpplint` diagnostics, including some false positives.
> 
> </details>


---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KHdeG1ti7FGiZUahFyA:-967573467)*
<!-- Sent from Reviewable.io -->

bradking commented:
> think about how this would be used to label the system methods (e.g. dynamics and output) as being a particular form w/ respect to their inputs. For instance, can we finally get rid of the isDirectFeedthrough and isTimeInvariant methods and replace them with a FunctionalForm label which contains that same information?

Actually I already have this prototyped.  Making this work was one of my design criteria for FunctionalForm.

In `drake/systems/test/testAffineSystem.cpp` I hacked the `testSizes` method with code like:

```c++
drake::FunctionalForm t = drake::FunctionalForm::Constant();
typename SysType::template StateVector<drake::FunctionalForm> x(num_states, 1);
x.setConstant(drake::FunctionalForm::Constant());
typename SysType::template InputVector<drake::FunctionalForm> u(num_inputs, 1);
for (std::size_t i = 0; i < num_inputs; ++i) {
  u(i) = drake::FunctionalForm::Linear({i});
}
std::cout << sys_ptr->output(t,x,u) << std::endl;  // prints "aff(0,1,2,3)" for each output
```



---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/core/functional_form.cc, line 24 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHWV00Og-18mZ5OuPhK:-KHeQW0wS6mseUcl-kPu:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/a2b288342742580790eae60e61b7fa6629819cca/drake/core/functional_form.cc#L24)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> Extract this repeated `8` into a constant.  Consider setting the constant to `static_cast<int>(kUndefined)+1` rather than explicit `8` (on the assumption that new enum values will be added before `kUndefined`).</details>
Done.

---

*[drake/core/functional_form.cc, line 500 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHWXYYkZkPzdYgOrenR:-KHeQWutL-fD8QwmciZM:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/a2b288342742580790eae60e61b7fa6629819cca/drake/core/functional_form.cc#L500)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> Is there a reason to use macros here instead of constants?</details>
Done.

---

*[drake/core/functional_form.h, line 33 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHWUdWO6Imu8kGRVbWn:-KHeQYJzqQb6sksp3VO8:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/a2b288342742580790eae60e61b7fa6629819cca/drake/core/functional_form.h#L33)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> This expression only goes up to quadratic.  You probably need a "..." somewhere in that example.</details>
Done.

---

*[drake/core/functional_form.h, line 75 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHdbrUZS1yr7JigVswM:-KHeT2WvOAKNZr1f4dyp:809531947) ([raw file](https://github.com/robotlocomotion/drake/blob/63c17cf988c78bc89e2d1e0242248d2bb8cb993d/drake/core/functional_form.h#L75)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> it looks like it is also true that f(x,3) would print aff(x).  that might be nice to add to this doc.</details>
Actually `f(x,3)` won't work because the template instantiation of `f<>` is ambiguous.  In practice these `f()` and `g()` calls will be the `eval()` or `output()` or `dynamics()` calls where all arguments are of the same scalar type.

What does work is `x * 3` and similar expressions, so I'll update the documentation for that part in the above section.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KHeQbmeJU-qzP8Jc20U:-79900728)*
<!-- Sent from Reviewable.io -->

bradking commented:
I've pushed a revised version of the topic to update documentation based on Russ's suggestion.  I also suppressed another cpplint false positive.

bradking commented:
I've pushed a revised version of the topic to rename the internal enumeration type `FunctionalForm::{Class => Form}`.  One day we may have to expose this enumeration publicly if clients need to `switch()` over the form.

bradking commented:
Review status: 3 of 5 files reviewed at latest revision, all discussions resolved.

---

*[drake/core/functional_form.h, line 304 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHdcSHYtkF9niCqcIdu:-KHem4wNDMt-2XvHBrF0:-1259844845) ([raw file](https://github.com/robotlocomotion/drake/blob/63c17cf988c78bc89e2d1e0242248d2bb8cb993d/drake/core/functional_form.h#L304)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> not required for this PR, but we should coordinate with @ggould-tri to decide if this functionality should be merged with the DecisionVariable logic in the constraint classes.</details>
While `FunctionalForm::Variable` and `DecisionVariable` may look similar at a glance I do not think they overlap.  Neither can play the role of the other, and I do not think there is a common abstraction that fits both use cases naturally.  However, we may be able to use `FunctionalForm::Variable` inside `DecisionVariable` as a replacement for the `std::string name_` field to offer more general identifiers.  I do not think we have a strong use case for this yet though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
"Actually I already have this prototyped. Making this work was one of my design criteria for FunctionalForm."

That is super cool - using FunctionalForm as an Eigen scalar will be a great way to getting insight into complex systems, and the code you pasted above has a straightforward System2 analogue.  I don't think it's a complete answer for how to label Systems, though, because it only gives you an answer for a particular code path, which could change in response to changing simulation state.  So, it seems likely the programmer will have to bear some of the burden of specifying (conservatively) the form/sparsity matrix.

Anyway, we should probably move this topic out of the margin notes of this PR, since it doesn't affect any of the code here.

<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> I've pushed a revised version of the topic to rename the internal enumeration type `FunctionalForm::{Class => Form}`.  One day we may have to expose this enumeration publicly if clients need to `switch()` over the form.
> 
> </details>


---

Review status: 3 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KHfH5_SGYoBpIaszijn:49335746)*
<!-- Sent from Reviewable.io -->

bradking commented:
I've pushed a revised version of the topic to squash in the Eigen alignment fix.

ggould-tri commented:
Okay, I've finally gotten through this.  Two quibbles about `#include`s; otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> I've pushed a revised version of the topic to squash in the Eigen alignment fix.
> 
> </details>


---

Reviewed 1 of 1 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/core/functional_form.cc, line 5 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHufYfRyLiVZRcHGuyh:-KHufYfRyLiVZRcHGuyi:-1342723089) ([raw file](https://github.com/robotlocomotion/drake/blob/62c804e17361cd2106df9a9e7b1cb492e45bbaf5/drake/core/functional_form.cc#L5)):*
> ```cpp
> #include <algorithm>
> #include <cassert>
> #include <iostream>
> ```

I believe the more specific `<ostream>` would work here

---

*[drake/core/functional_form.h, line 304 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHdcSHYtkF9niCqcIdu:-KHu_Gd5qKD4zgckVbJr:1286662007) ([raw file](https://github.com/robotlocomotion/drake/blob/63c17cf988c78bc89e2d1e0242248d2bb8cb993d/drake/core/functional_form.h#L304)):*
<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> While `FunctionalForm::Variable` and `DecisionVariable` may look similar at a glance I do not think they overlap.  Neither can play the role of the other, and I do not think there is a common abstraction that fits both use cases naturally.  However, we may be able to use `FunctionalForm::Variable` inside `DecisionVariable` as a replacement for the `std::string name_` field to offer more general identifiers.  I do not think we have a strong use case for this yet though.
> </details>
I believe that this Variable representation is unconditionally superior to the one in Polynomial (unless we somewhere rely on Polynomial::VarType compatibility with the spotless implementation, but I believe the earlier bug in it proved we cannot be relying on that in practice).  So I'm inclined to unify those at some arbitrary point in the future (definitely my problem, not yours!)

The argument for unifying this with decision variables is simply interface tidiness.  We can cross that bridge when we come to it.

---

*[drake/core/functional_form.h, line 695 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHubgrEKURmokRvOfKn:-KHubgrEKURmokRvOfKo:1274459955) ([raw file](https://github.com/robotlocomotion/drake/blob/62c804e17361cd2106df9a9e7b1cb492e45bbaf5/drake/core/functional_form.h#L695)):*
> ```objc
>   return lhs /= FunctionalForm(rhs);
> }
> ```

That is some deeply heroic template magic right there.  I learned a lot from reading it.

---

*[drake/core/test/functional_form_test.cc, line 6 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHuckkTLi9AKD8kxrvD:-KHuckkTLi9AKD8kxrvE:2072071696) ([raw file](https://github.com/robotlocomotion/drake/blob/62c804e17361cd2106df9a9e7b1cb492e45bbaf5/drake/core/test/functional_form_test.cc#L6)):*
> ```cpp
> #include <Eigen/Dense>
> 
> #include <iostream>
> ```

I believe this may be unused.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KHuebVM-bBkPjeGUW_s:-1614059142)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Reviewed 1 of 1 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322)*
<!-- Sent from Reviewable.io -->

bradking commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/core/functional_form.cc, line 5 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHufYfRyLiVZRcHGuyh:-KHvFcKc77LulsulA6fl:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/62c804e17361cd2106df9a9e7b1cb492e45bbaf5/drake/core/functional_form.cc#L5)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> I believe the more specific `<ostream>` would work here</details>
Done.

---

*[drake/core/test/functional_form_test.cc, line 6 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KHuckkTLi9AKD8kxrvD:-KHvFgHdLN_tQXSa4uPm:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/62c804e17361cd2106df9a9e7b1cb492e45bbaf5/drake/core/test/functional_form_test.cc#L6)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> I believe this may be unused.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322)*
<!-- Sent from Reviewable.io -->

bradking commented:
I've pushed a revised version of the topic to address @ggould-tri's comments on `#include` tidiness.

bradking commented:
@ggould-tri I think this is ready for final review and integration now.

ggould-tri commented:
I'm only one reviewer, so I shouldn't be both feature and platform review.  How about:
+@david-german-tri (platform)

<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> @ggould-tri I think this is ready for final review and integration now.
> 
> </details>


---

Reviewed 2 of 2 files at r9.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KHzlfk5O5FStfti-kd-:1834316539)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Woohoo!  I have some questions, all relatively minor.

<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> I'm only one reviewer, so I shouldn't be both feature and platform review.  How about:
> +@david-german-tri (platform)
> </details>


---

Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[drake/core/functional_form.cc, line 121 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-BeY27aW5VOmzzTcJ:-KI-BeY27aW5VOmzzTcK:-1063576180) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.cc#L121)):*
> ```cpp
> FunctionalForm operator+(FunctionalForm const& l, FunctionalForm const& r) {
>   FunctionalForm::Form f =
>       FunctionalForm::Internal::kAdd[int(l.form_)][int(r.form_)];
> ```

nit: C-style casts are prohibited in the style guide.  https://google.github.io/styleguide/cppguide.html#Casting

---

*[drake/core/functional_form.h, line 121 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-9RF7-yubzNIAim5T:-KI-9RF7-yubzNIAim5U:1049822432) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L121)):*
> ```objc
>   static FunctionalForm Constant();
> 
>   /** Return a \ref linear form of one or more variables. */
> ```

As a user, I'd be surprised to find out that `v` got `std::move`d after calling this function.  Why not just copy?  The style-guide restrictions are on rvalue references are wishy-washy and I'm willing to bend them, but I don't understand the benefit in this PR.

https://google.github.io/styleguide/cppguide.html#Rvalue_references

---

*[drake/core/functional_form.h, line 308 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-A85XaeQ3oORBY_tH:-KI-A85XaeQ3oORBY_tI:767688717) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L308)):*
> ```objc
>    * The contained set is immutable.
>    */
>   class DRAKECORE_EXPORT Variables {
> ```

What's the benefit of nesting this class?

---

*[drake/core/functional_form.h, line 366 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-6eiQ8v2n8H0nww8s:-KI-6eiQ8v2n8H0nww8t:-1280128995) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L366)):*
> ```objc
> 
>  private:
> #if !defined(DRAKE_DOXYGEN_CXX)
> ```

What's the reason for carving this stuff out of doxygen?

---

*[drake/core/functional_form.h, line 706 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-8rjKfcN4B-xt7K5t:-KI-8rjKfcN4B-xt7K5u:1014375585) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L706)):*
> ```objc
> struct NumTraits<drake::FunctionalForm> {
>   enum {
>     IsInteger = 1,
> ```

really?  `cos`, `sqrt`, and so on implemented, right?

---

*[drake/core/functional_form.h, line 720 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-8OyIHf6UVUci8XZd:-KI-8OyIHf6UVUci8XZe:-1488273146) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L720)):*
> ```objc
>   static inline Real dummy_precision() { return drake::FunctionalForm(); }
> #if 0
>   // Enable these when we encounter a need for them.
> ```

is it useful to have them here at all?

---

*[drake/core/functional_form.h, line 721 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-8Bg5GE7k-PhNJfz6:-KI-8Bg6l382A-ayFBnJ:-1375742049) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L721)):*
> ```objc
> #if 0
>   // Enable these when we encounter a need for them.
>   static inline Real epsilmn() { return FunctionalForm(); }
> ```

is this a typo?  if not, what does it mean?

---

*[drake/core/test/functional_form_test.cc, line 5 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-Hzuevqtdv6w_hqET:-KI-Hzuevqtdv6w_hqEU:1521522744) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/test/functional_form_test.cc#L5)):*
> ```cpp
> #include <Eigen/Core>
> 
> #include <sstream>
> ```

nit: order of includes

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KI-IbuDpSFFSJHGRFHQ:-1623155402)*
<!-- Sent from Reviewable.io -->

bradking commented:
Thanks for the comments.  I'll push a revised version of the topic to address them.

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Woohoo!  I have some questions, all relatively minor.
</details>


---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/core/functional_form.cc, line 121 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-BeY27aW5VOmzzTcJ:-KI38m_MWFqClxB8932D:2137662548) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.cc#L121)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> nit: C-style casts are prohibited in the style guide.  https://google.github.io/styleguide/cppguide.html#Casting
</details>
In the next revision I've replaced this with a more readable `IndexOf(x)` expression.

FYI, `int(x)` is not actually a C-style cast.  It is a call to the `int` type "constructor" which for primitive types is exactly the behavior of `static_cast<int>(x)`.  However, I see the Google style guide discourages the style anyway for some reason.

---

*[drake/core/functional_form.h, line 121 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-9RF7-yubzNIAim5T:-KI3BFqNuLiu9nHV0b2y:1234603463) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L121)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> As a user, I'd be surprised to find out that `v` got `std::move`d after calling this function.  Why not just copy?  The style-guide restrictions are on rvalue references are wishy-washy and I'm willing to bend them, but I don't understand the benefit in this PR.
> 
> https://google.github.io/styleguide/cppguide.html#Rvalue_references
</details>
These functions take `Variables v` by value.  Each implementation then moves from the value it already owns.  This is value semantics for the caller.

FYI, functions that take rvalue references cannot silently move out of a caller's lvalue because they won't bind.  The caller needs to cast lvalue-to-rvalue with an explicit `std::move`.

---

*[drake/core/functional_form.h, line 308 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-A85XaeQ3oORBY_tH:-KI3Bk3peU6obDaxBP_x:1260311116) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L308)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> What's the benefit of nesting this class?
</details>
Its role is very specific to `FunctionalForm` so it makes sense as a nested type.  I would prefer to declare inside and then define outside (as is done for `Variable`), but the `Variables` class definition has to be nested because we use it as a data member of `FunctionalForm` and so must be a complete type early enough.

---

*[drake/core/functional_form.h, line 366 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-6eiQ8v2n8H0nww8s:-KI3J-m-zsRwzLTZNrgg:547261462) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L366)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> What's the reason for carving this stuff out of doxygen?
</details>
Doxygen was incorrectly documenting some members as public.  For the next revision I've made the exclusions more specific and added corresponding comments.

---

*[drake/core/functional_form.h, line 706 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-8rjKfcN4B-xt7K5t:-KI3CAGZCGoLZNNgCx9L:2140251033) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L706)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> really?  `cos`, `sqrt`, and so on implemented, right?
</details>
I'm not intimately familiar with what Eigen does with this trait, but both its set of allowed values and its equality comparison operation are (unsigned-)integer-like (e.g. no epsilon is allowed as with `double`).

---

*[drake/core/functional_form.h, line 720 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-8OyIHf6UVUci8XZd:-KI3E5NWdRWUPXPVb8jd:1887919534) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L720)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> is it useful to have them here at all?
</details>
Oops, this block was left from development when I was still determining what traits are needed to support Eigen for the operations we need.  I've dropped this block for the next revision.

---

*[drake/core/functional_form.h, line 721 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-8Bg5GE7k-PhNJfz6:-KI3EGbL4JoKSRIWhnV6:-1123510093) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L721)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> is this a typo?  if not, what does it mean?
</details>
  I've dropped this block for the next revision.

---

*[drake/core/test/functional_form_test.cc, line 5 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-Hzuevqtdv6w_hqET:-KI3EgDy-PsjQbM0hv66:-185442994) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/test/functional_form_test.cc#L5)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> nit: order of includes
> 
> https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes
</details>
Fixed for next revision.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KI3JOlZOQDImMv0XLRP:-489772119)*
<!-- Sent from Reviewable.io -->

bradking commented:
I've pushed a revised version of the topic to address @david-german-tri's comments.

bradking commented:
It seems Reviewable does not allow a discussion to be marked resolved after a comment has been published without adding yet another comment.

<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> I've pushed a revised version of the topic to address @david-german-tri's comments.
</details>


---

Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/core/functional_form.cc, line 121 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-BeY27aW5VOmzzTcJ:-KI3uXJQE_OsE8svCl-l:-864509007) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.cc#L121)):*
<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> In the next revision I've replaced this with a more readable `IndexOf(x)` expression.
> 
> FYI, `int(x)` is not actually a C-style cast.  It is a call to the `int` type "constructor" which for primitive types is exactly the behavior of `static_cast<int>(x)`.  However, I see the Google style guide discourages the style anyway for some reason.
> 
</details>
Resolved.

---

*[drake/core/functional_form.h, line 121 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-9RF7-yubzNIAim5T:-KI3ugQZxul6634MNNH3:-864509007) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L121)):*
<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> These functions take `Variables v` by value.  Each implementation then moves from the value it already owns.  This is value semantics for the caller.
> 
> FYI, functions that take rvalue references cannot silently move out of a caller's lvalue because they won't bind.  The caller needs to cast lvalue-to-rvalue with an explicit `std::move`.
</details>
Resolved.

---

*[drake/core/functional_form.h, line 308 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-A85XaeQ3oORBY_tH:-KI3rVCMXFYYCbErqtJf:-864509007) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L308)):*
<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> Its role is very specific to `FunctionalForm` so it makes sense as a nested type.  I would prefer to declare inside and then define outside (as is done for `Variable`), but the `Variables` class definition has to be nested because we use it as a data member of `FunctionalForm` and so must be a complete type early enough.
</details>
Resolved.

---

*[drake/core/functional_form.h, line 366 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-6eiQ8v2n8H0nww8s:-KI3rU0HkEKvVp3o0Z_g:-864509007) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L366)):*
<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> Doxygen was incorrectly documenting some members as public.  For the next revision I've made the exclusions more specific and added corresponding comments.
</details>
Resolved.

---

*[drake/core/functional_form.h, line 706 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-8rjKfcN4B-xt7K5t:-KI3utK4BBRUASS41jvY:-864509007) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L706)):*
<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> I'm not intimately familiar with what Eigen does with this trait, but both its set of allowed values and its equality comparison operation are (unsigned-)integer-like (e.g. no epsilon is allowed as with `double`).
</details>
Resolved.

---

*[drake/core/functional_form.h, line 720 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-8OyIHf6UVUci8XZd:-KI3upblgi4xnAijpeb4:-864509007) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L720)):*
<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> Oops, this block was left from development when I was still determining what traits are needed to support Eigen for the operations we need.  I've dropped this block for the next revision.
</details>
Resolved.

---

*[drake/core/functional_form.h, line 721 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-8Bg5GE7k-PhNJfz6:-KI3uoppKkixiDgIpFh-:-864509007) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/functional_form.h#L721)):*
<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
>   I've dropped this block for the next revision.
</details>
Resolved.

---

*[drake/core/test/functional_form_test.cc, line 5 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-KI-Hzuevqtdv6w_hqET:-KI3uvYcrQp4foBYkVNd:-864509007) ([raw file](https://github.com/robotlocomotion/drake/blob/27ddb289e6968638f608dc3adf910348e6e76af9/drake/core/test/functional_form_test.cc#L5)):*
<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> Fixed for next revision.
</details>
Resolved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KI3uxefjo2LCLsMsU0O:69662624)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, bradking (Brad King) wrote…</i></summary>
> It seems Reviewable does not allow a discussion to be marked resolved after a comment has been published without adding yet another comment.
</details>


---

Review status: 2 of 5 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KI3yXfw6ulCaSpJPjVS:1417088721)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Should we go ahead and merge this?

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> :lgtm:
</details>


---

Review status: 2 of 5 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2322#-:-KI3yYlUHshaZUXoA5z3:1170337059)*
<!-- Sent from Reviewable.io -->

bradking commented:
> Should we go ahead and merge this?

Yes, I think it is ready now.

jwnimmer-tri commented:
Using override-merge due to out-of-date-with-master.

jwnimmer-tri commented:
For the record...

The final jenkins run on this PR was around 1pm today.  Sometime after that, PRs switched from the studio generator to the ninja generator.  The unit test here is failing when run under windows with ninja; it passes with the studio generator (as seen in Continuous).

The effect is that new PRs show a test failure in their pre-submit builds, because master is "broken", even though the test failure only occurs on an unsupported configuration.

jwnimmer-tri commented:
Update: it looks like the studio debug continuous are also showing up as broken now; only release studio is working.  Given that a supported configuration is now showing up as broken, a revert is probably in order.

bradking commented:
> it looks like the studio debug continuous are also showing up as broken now; only release studio is working.

Recorded as issue #2398.  Will fix ASAP.

bradking commented:
See PR #2400.
